snuba/datasets/plans/sliced_storage.py

```diff
@@ -81,12 +120,13 @@ def select_cluster(
         assert len(org_ids) == 1
         org_id = org_ids.pop()
 
-        slice_id = map_logical_partition_to_slice(
-            self.storage, map_org_id_to_logical_partition(org_id)
-        )
-        cluster = get_cluster(self.storage_set, slice_id)
-
-        return cluster
+        logical_partition = map_org_id_to_logical_partition(org_id)
+        if _should_use_mega_cluster(self.storage_set, logical_partition):
+            return get_cluster(self.storage_set)
```

Comment:
I was sort of assuming that the storage set key for the "mega-cluster" would be distinct from the sliced cluster, though it's not a requirement. Presumably the run-time config used to define `_should_use_mega_cluster` could store that storage set key in the value.

Did you consider this and dismiss it or just assume that we would want the storage set key to remain consistent whether using the mega-cluster or the normal sliced cluster?

Reply:
I was also wondering about this. Currently, `settings.CLUSTERS` is only for non-sliced storage set keys, `settings.SLICED_CLUSTERS` only for sliced storage set keys. We can't have a storage set key be present in both.

Reply:
I considered using a separate storage_set but decided against it and using a simpler approach. My thought process was the following.

The way to access a clickhouse cluster is via `storage_set`. Example [here](https://github.com/getsentry/snuba/blob/master/snuba/settings/__init__.py#L61-L77). Lets assume mega cluster access should happen via a separate `storage_set`. In that case we would want a generic_metrics [entity](https://github.com/getsentry/snuba/blob/master/snuba/datasets/configuration/generic_metrics/entities/distributions.yaml#L3) to map to a mega cluster `storage_set`. Today it links with a `storage_set` from [here](https://github.com/getsentry/snuba/blob/master/snuba/datasets/configuration/generic_metrics/entities/distributions.yaml#L69) and then [here](https://github.com/getsentry/snuba/blob/master/snuba/datasets/configuration/generic_metrics/storages/distributions.yaml#L7)

To enable that lets see what would need to happen. `storage_set` is defined in the yaml files like [this](https://github.com/getsentry/snuba/blob/master/snuba/datasets/configuration/generic_metrics/storages/distributions.yaml#L7). That code is used as part of building the [storage schema](https://github.com/getsentry/snuba/blob/master/snuba/datasets/configuration/storage_builder.py#L87) and [readable storage](https://github.com/getsentry/snuba/blob/master/snuba/datasets/configuration/storage_builder.py#L95). `ReadableTableStorage` is linked with an `Entity`. An `Entity` can only have one `ReadableTableStorage` as shown [here](https://github.com/getsentry/snuba/blob/master/snuba/datasets/pluggable_entity.py#L44). We don't want to specify multiple `ReadableTableStorage` for a single `Entity`. That means defining a new `storage_set` would imply defining a new `Entity` as well which we do not want in this case. And it would also imply keeping schema `columns` same across different datasets and migrations consistent across different datasets. Which is another big headache. Also, the new storage_key is probably needed only when adding new slices. It has no purpose after all data has been migrated.

So I would like to simplify the problem. And the simplification is that somehow an API call needs to be routed to the mega-cluster. mega-cluster is just another host/port definition which needs to be looked up. If we want the mega-cluster definition to not be in settings.CLUSTER, a workaround I can think of is defining a reserved slice number (say 10000) which would provide access to the mega-cluster via settings.SLICED_CLUSTERS.

Reply:
That argument makes sense to me. 

@ayirr7, do you see an inherent conflict between Nikhar's approach and the work you've been doing? You say 

> We can't have a storage set key be present in both.

So is this going to fail to start `snuba api` or `snuba consumer --storage=<sliced storage>` during validation or is this just violating an assumption?

Reply:
That also makes sense to me. We can keep it in `settings.CLUSTERS` and I don't think any validation would break because of this. I don't think it makes sense to define any mega cluster in `settings.SLICED_CLUSTERS` because the mega cluster can't be sliced. I think it'd be helpful to add this in docs or comments in the settings file (the mega cluster for a sliced storage set would be defined in `settings.CLUSTERS` and not in `SLICED_CLUSTERS`).